File: buddy_viewer/viewer.ts
#!/usr/bin/env bun
/**
 * Animated companion viewer with rich terminal UI.
 * Built with Bun + chalk for proper UTF-8 and color handling.
 *
 * Controls:
 *   ←/a  Previous    →/d  Next    ↑  Jump 10 back    ↓  Jump 10 forward
 *   w/s  Species filter    q/e  Rarity filter    x  Exit
 *   --core  Browse only featured companions
 */

import chalk from "chalk";
import { readFileSync, readdirSync, existsSync } from "fs";
import { join, dirname, basename } from "path";

// ── Config ──
const REPO = "/home/sprite/claude-buddies";
const FRAME_DELAY = 800; // ms
const CARD_WIDTH = 68;
const LEFT_PAD = "          "; // 10 chars

const ALL_SPECIES = ["all", "axolotl", "blob", "cactus", "capybara", "cat", "chonk", "dragon", "duck", "ghost", "goose", "mushroom", "octopus", "owl", "penguin", "rabbit", "robot", "snail", "turtle"];
const ALL_RARITIES = ["all", "common", "uncommon", "rare", "epic", "legendary"];
const CORE_LIST = ["anarathil", "eldurion", "faelindor", "ithrandur", "morgrath", "thistlewing", "thunderthistl", "vyrenth"];

const RARITY_STARS: Record<string, string> = {
  common: "★", uncommon: "★★", rare: "★★★", epic: "★★★★", legendary: "★★★★★",
};

// ── Parse args ──
const args = process.argv.slice(2);
const coreOnly = args.includes("--core");
const startAt = args.find((_, i) => args[i - 1] === "--start") ?? "";

// ── Colors ──
const gold = chalk.hex("#FFC107");
const orange = chalk.hex("#D7775B");
const dim = chalk.hex("#888888");
const cyan = chalk.hex("#82AADC");
const green = chalk.hex("#93C882");
const red = chalk.hex("#FF6B80");
const white = chalk.bold.white;
const CLR = "\x1b[K"; // clear to end of line — prevents ghost chars from previous frames

// Rarity color themes: [border, accent, sprite]
const RARITY_THEMES: Record<string, { border: chalk.Chalk; accent: chalk.Chalk; sprite: chalk.Chalk }> = {
  common:    { border: chalk.hex("#9E9E9E"), accent: chalk.hex("#BDBDBD"), sprite: chalk.hex("#B0B0B0") },
  uncommon:  { border: chalk.hex("#4CAF50"), accent: chalk.hex("#81C784"), sprite: chalk.hex("#A5D6A7") },
  rare:      { border: chalk.hex("#2196F3"), accent: chalk.hex("#64B5F6"), sprite: chalk.hex("#90CAF9") },
  epic:      { border: chalk.hex("#9C27B0"), accent: chalk.hex("#CE93D8"), sprite: chalk.hex("#E1BEE7") },
  legendary: { border: chalk.hex("#D7775B"), accent: chalk.hex("#FFC107"), sprite: chalk.hex("#FFC107") },
};

// ── Load companions ──
interface Companion {
  path: string;
  folder: string;
  name: string;
  species: string;
  rarity: string;
  eye: string;
  hat: string;
  shiny: boolean;
  personality: string;
  peakStat: string;
  dumpStat: string;
  stats: Record<string, number>;
  spriteFile: string;
}

function findAllBuddyJsons(): string[] {
  const results: string[] = [];
  function walk(dir: string, depth = 0) {
    if (depth > 5) return;
    try {
      for (const entry of readdirSync(dir, { withFileTypes: true })) {
        const full = join(dir, entry.name);
        if (entry.isFile() && entry.name === "buddy.json") results.push(full);
        else if (entry.isDirectory() && entry.name !== "info" && entry.name !== "node_modules") walk(full, depth + 1);
      }
    } catch {}
  }
  walk(join(REPO, "companions"));
  return results.sort();
}

function loadCompanion(buddyPath: string): Companion {
  const data = JSON.parse(readFileSync(buddyPath, "utf-8"));
  const dir = dirname(buddyPath);
  return {
    path: buddyPath,
    folder: basename(dir),
    name: data.name,
    species: data.bones.species,
    rarity: data.bones.rarity,
    eye: data.bones.eye,
    hat: data.bones.hat,
    shiny: data.bones.shiny,
    personality: data.personality,
    peakStat: data.peakStat ?? "?",
    dumpStat: data.dumpStat ?? "?",
    stats: data.bones.stats,
    spriteFile: join(dir, "sprite.txt"),
  };
}

console.log("Loading companions...");
const allBuddyPaths = findAllBuddyJsons();
let allCompanions: Companion[];

if (coreOnly) {
  allCompanions = CORE_LIST
    .map((c) => join(REPO, "companions", c, "buddy.json"))
    .filter(existsSync)
    .map(loadCompanion);
} else {
  allCompanions = allBuddyPaths.map(loadCompanion);
}
console.log(`Loaded ${allCompanions.length} companions.`);

// ── State ──
let speciesIdx = 0;
let rarityIdx = 0;
let filtered: Companion[] = [];
let index = 0;
let frame = 0;

function rebuildFilter() {
  const sp = ALL_SPECIES[speciesIdx];
  const ra = ALL_RARITIES[rarityIdx];
  filtered = allCompanions.filter((c) => {
    if (sp !== "all" && c.species !== sp) return false;
    if (ra !== "all" && c.rarity !== ra) return false;
    return true;
  });
  index = 0;
}
rebuildFilter();

// Find start position
if (startAt) {
  const idx = filtered.findIndex((c) => c.folder === startAt);
  if (idx >= 0) index = idx;
}

// ── Sprite frames ──
function getFrame(comp: Companion, frameNum: number, sleeping: boolean): string[] {
  if (!existsSync(comp.spriteFile)) return ["  (no sprite)"];
  const text = readFileSync(comp.spriteFile, "utf-8");
  const label = sleeping ? "Sleeping" : `Frame ${frameNum}`;
  const regex = new RegExp(`--- ${label}[^-]*---\\n([\\s\\S]*?)(?=\\n---|\\n*$)`);
  const match = text.match(regex);
  if (!match) return ["  (frame not found)"];
  return match[1].split("\n").filter((l) => l.length > 0).slice(0, 5);
}

// ── Rendering ──
function statBar(val: number): string {
  const filled = Math.floor(val / 5);
  const empty = 20 - filled;
  return white("█".repeat(filled)) + dim("░".repeat(empty));
}

function padRight(str: string, width: number): string {
  // Visual width (not byte length)
  const visLen = [...str].length;
  const pad = Math.max(0, width - visLen);
  return str + " ".repeat(pad);
}

function hline(): string {
  return "─".repeat(CARD_WIDTH);
}

function blankLine(): string {
  return " ".repeat(CARD_WIDTH);
}



function cardLine(content: string, visWidth: number): string {
  const pad = Math.max(0, CARD_WIDTH - 1 - visWidth);
  return `${orange("│")} ${content}${" ".repeat(pad)}${orange("│")}${CLR}`;
}

function renderCard(comp: Companion, frameLines: string[], sleeping: boolean): string {
  const lines: string[] = [];
  const stars = RARITY_STARS[comp.rarity] ?? "?";
  const speciesUpper = comp.species.toUpperCase();
  const rarityUpper = comp.rarity.toUpperCase();
  const shinyTag = comp.shiny ? " ✨" : "";
  const totalFiltered = filtered.length;
  const navPos = `${index + 1}/${totalFiltered}`;

  // Rarity-based colors
  const theme = RARITY_THEMES[comp.rarity] ?? RARITY_THEMES.legendary;
  const b = theme.border;  // border color
  const a = theme.accent;  // accent (headers, stats, stars)
  const s = theme.sprite;  // sprite color

  // Top border
  lines.push(`${LEFT_PAD}${b("╭" + hline() + "╮")}`);

  // Header: rarity + species
  const headerLeft = `${stars}  ${rarityUpper}`;
  const headerGap = CARD_WIDTH - 2 - headerLeft.length - speciesUpper.length;
  lines.push(`${LEFT_PAD}${b("│")} ${a(headerLeft)}${" ".repeat(Math.max(0, headerGap))}${cyan(speciesUpper)} ${b("│")}${CLR}`);

  // Blank
  lines.push(`${LEFT_PAD}${b("│")}${blankLine()}${b("│")}${CLR}`);

  // Sprite
  for (let i = 0; i < 5; i++) {
    const spriteLine = frameLines[i] ?? "";
    const visLen = [...spriteLine].length;
    const pad = Math.max(0, CARD_WIDTH - 1 - visLen);
    lines.push(`${LEFT_PAD}${b("│")} ${s(spriteLine)}${" ".repeat(pad)}${b("│")}${CLR}`);
  }

  // Blank
  lines.push(`${LEFT_PAD}${b("│")}${blankLine()}${b("│")}${CLR}`);

  // Name (✨ is 2 terminal columns but 1 JS char — subtract extra for wide chars)
  const nameDisplay = `${comp.name}${shinyTag}`;
  const wideChars = [...nameDisplay].filter(c => c === "✨").length;
  const namePad = Math.max(0, CARD_WIDTH - 1 - [...nameDisplay].length - wideChars);
  lines.push(`${LEFT_PAD}${b("│")} ${white(nameDisplay)}${" ".repeat(namePad)}${b("│")}${CLR}`);

  // Blank
  lines.push(`${LEFT_PAD}${b("│")}${blankLine()}${b("│")}${CLR}`);

  // Personality (3 lines max, word-wrapped)
  const maxPW = CARD_WIDTH - 4;
  const words = comp.personality.split(" ");
  const pLines: string[] = [];
  let current = "";
  for (const word of words) {
    if (current.length + word.length + 1 > maxPW) {
      pLines.push(current);
      current = word;
    } else {
      current = current ? current + " " + word : word;
    }
  }
  if (current) pLines.push(current);
  for (const pl of pLines.slice(0, 3)) {
    const pPad = Math.max(0, CARD_WIDTH - 1 - pl.length);
    lines.push(`${LEFT_PAD}${b("│")} ${dim(pl)}${" ".repeat(pPad)}${b("│")}${CLR}`);
  }

  // Blank
  lines.push(`${LEFT_PAD}${b("│")}${blankLine()}${b("│")}${CLR}`);

  // Stats
  const statNames = ["DEBUGGING", "PATIENCE", "CHAOS", "WISDOM", "SNARK"] as const;
  for (const stat of statNames) {
    const val = comp.stats[stat] ?? 0;
    const bar = statBar(val);
    let suffix = "";
    if (stat === comp.peakStat) suffix = " (peak)";
    else if (stat === comp.dumpStat) suffix = " (dump)";
    // Visual width: 2(pad) + 10(name) + 1(sp) + 20(bar) + 1(sp) + 3(val) + suffix + 2(pad) = 39 + suffix
    const visWidth = 2 + 10 + 1 + 20 + 1 + 3 + suffix.length + 2;
    const rightPad = Math.max(0, CARD_WIDTH - visWidth);
    const statLine = `  ${a(stat.padEnd(10))} ${bar} ${dim(String(val).padStart(3))}${dim(suffix)}${" ".repeat(rightPad)}  `;
    lines.push(`${LEFT_PAD}${b("│")}${statLine}${b("│")}${CLR}`);
  }

  // Blank
  lines.push(`${LEFT_PAD}${b("│")}${blankLine()}${b("│")}${CLR}`);

  // Status
  if (sleeping) {
    const statusPad = Math.max(0, CARD_WIDTH - 17);
    lines.push(`${LEFT_PAD}${b("│")}  ${dim("💤 sleeping...")}${" ".repeat(statusPad)}${b("│")}${CLR}`);
  } else {
    const statusPad = Math.max(0, CARD_WIDTH - 11);
    lines.push(`${LEFT_PAD}${b("│")}  ${green("● online")}${" ".repeat(statusPad)}${b("│")}${CLR}`);
  }

  // Bottom border
  lines.push(`${LEFT_PAD}${b("╰" + hline() + "╯")}`);
  lines.push("");

  // Navigation footer (clear to end of line to prevent ghost text from previous frames)
  
  lines.push(`${LEFT_PAD}  ${dim("a/← prev")}  ${white(`[ ${navPos} ]`)}  ${dim("d/→ next")}  ${dim("↑↓ jump")}  ${dim("x exit")}${CLR}`);
  if (!coreOnly) {
    lines.push(`${LEFT_PAD}  ${dim("Species:")} ${cyan(ALL_SPECIES[speciesIdx])} ${dim("(w/s)")}  ${dim("Rarity:")} ${cyan(ALL_RARITIES[rarityIdx])} ${dim("(q/e)")}  ${dim(comp.folder)}${CLR}`);
  } else {
    lines.push(`${LEFT_PAD}  ${dim("--core mode")}  ${dim(comp.folder)}${CLR}`);
  }

  return lines.join("\n");
}

function renderEmpty(): string {
  const lines: string[] = [];
  lines.push(`${LEFT_PAD}${orange("╭" + hline() + "╮")}`);
  lines.push(`${LEFT_PAD}${orange("│")}${blankLine()}${orange("│")}${CLR}`);
  const msg = "No companions match filters";
  const pad = Math.floor((CARD_WIDTH - msg.length) / 2);
  lines.push(`${LEFT_PAD}${orange("│")}${" ".repeat(pad)}${red(msg)}${" ".repeat(CARD_WIDTH - pad - msg.length)}${orange("│")}${CLR}`);
  lines.push(`${LEFT_PAD}${orange("│")}${blankLine()}${orange("│")}${CLR}`);
  lines.push(`${LEFT_PAD}${orange("╰" + hline() + "╯")}`);
  lines.push("");
  lines.push(`${LEFT_PAD}  ${dim("Species:")} ${white(ALL_SPECIES[speciesIdx])} ${dim("(w/s)")}  ${dim("Rarity:")} ${white(ALL_RARITIES[rarityIdx])} ${dim("(q/e)")}  ${dim("x exit")}\x1b[K`);
  return lines.join("\n");
}

// ── Terminal control ──
function hideCursor() { process.stdout.write("\x1b[?25l"); }
function showCursor() { process.stdout.write("\x1b[?25h"); }
function moveTo(row: number, col: number) { process.stdout.write(`\x1b[${row};${col}H`); }
function clearScreen() { process.stdout.write("\x1b[2J"); moveTo(1, 1); }

// ── Input handling ──
function setupRawInput() {
  if (process.stdin.isTTY) {
    process.stdin.setRawMode(true);
  }
  process.stdin.resume();
  process.stdin.setEncoding("utf-8");
}

let pendingKey = "";

process.stdin.on("data", (key: string) => {
  pendingKey = key;
});

function consumeKey(): string {
  const k = pendingKey;
  pendingKey = "";
  return k;
}

// ── Animation loop ──
function getFrameData(): { lines: string[]; sleeping: boolean } {
  const comp = filtered[index];
  if (!comp) return { lines: [], sleeping: false };

  switch (frame) {
    case 0: case 1: case 3: case 4:
      return { lines: getFrame(comp, frame % 2, false), sleeping: false };
    case 2:
      return { lines: getFrame(comp, 2, false), sleeping: false };
    case 5:
      return { lines: getFrame(comp, 0, true), sleeping: true };
    default:
      return { lines: getFrame(comp, 0, false), sleeping: false };
  }
}

function handleInput(key: string): boolean {
  const total = filtered.length;

  switch (key) {
    case "\x1b[D": case "a": case "A": // Left
      if (total > 0) index = (index - 1 + total) % total;
      frame = 0;
      return true;
    case "\x1b[C": case "d": case "D": // Right
      if (total > 0) index = (index + 1) % total;
      frame = 0;
      return true;
    case "\x1b[A": // Up — jump 10
      if (total > 0) index = (index - 10 + total) % total;
      frame = 0;
      return true;
    case "\x1b[B": // Down — jump 10
      if (total > 0) index = (index + 10) % total;
      frame = 0;
      return true;
    case "e": case "E": // Rarity up
      if (!coreOnly) {
        rarityIdx = (rarityIdx + 1) % ALL_RARITIES.length;
        rebuildFilter();
      }
      return true;
    case "q": // Rarity down
      if (!coreOnly) {
        rarityIdx = (rarityIdx - 1 + ALL_RARITIES.length) % ALL_RARITIES.length;
        rebuildFilter();
      }
      return true;
    case "s": case "S": // Species next
      if (!coreOnly) {
        speciesIdx = (speciesIdx + 1) % ALL_SPECIES.length;
        rebuildFilter();
      }
      return true;
    case "w": case "W": // Species prev
      if (!coreOnly) {
        speciesIdx = (speciesIdx - 1 + ALL_SPECIES.length) % ALL_SPECIES.length;
        rebuildFilter();
      }
      return true;
    case "x": case "X": case "\x03": // Exit (x or Ctrl+C)
      return false;
    default:
      return true;
  }
}

async function main() {
  setupRawInput();
  hideCursor();
  clearScreen();

  const cleanup = () => {
    showCursor();
    clearScreen();
    const comp = filtered[index];
    console.log(`Goodbye from ${comp?.name ?? "companion"}!`);
    process.exit(0);
  };

  process.on("SIGINT", cleanup);
  process.on("SIGTERM", cleanup);

  while (true) {
    moveTo(1, 1);

    if (filtered.length === 0) {
      process.stdout.write(renderEmpty());
    } else {
      const { lines, sleeping } = getFrameData();
      const comp = filtered[index];
      process.stdout.write(renderCard(comp, lines, sleeping));
    }

    // Wait for frame delay, checking for input
    await new Promise<void>((resolve) => setTimeout(resolve, FRAME_DELAY));

    const key = consumeKey();
    if (key) {
      const cont = handleInput(key);
      if (!cont) {
        cleanup();
        return;
      }
    }

    frame = (frame + 1) % 6;
  }
}

main().catch(console.error);

File: buddy_viewer/viewer.py
#!/usr/bin/env python3
"""
Animated companion viewer — pure Python, no external libraries.

Controls:
  ← / → or a/d   Browse companions (prev / next)
  ↑ / ↓           Jump 10 companions
  e / q            Cycle rarity filter
  w / s            Cycle species filter
  x                Exit

Usage:
  python3 viewer.py                     # browse all
  python3 viewer.py --core              # core companions only
  python3 viewer.py --start eldurion    # start at companion
  python3 viewer.py --delay 0.5         # faster animation
"""

import json
import os
import sys
import time
import select
import termios
import tty
from pathlib import Path

# ── Config ──
REPO = Path("/home/sprite/claude-buddies")
CARD_WIDTH = 68
LEFT_PAD = "          "  # 10 chars
FRAME_DELAY = 0.8
CLR = "\033[K"

# ── ANSI color helpers (no libraries needed) ──
def rgb(r, g, b):
    """Return a function that wraps text in RGB color."""
    def colorize(text):
        return f"\033[38;2;{r};{g};{b}m{text}\033[0m"
    return colorize

def bold_white(text):
    return f"\033[1;37m{text}\033[0m"

def dim_text(text):
    return f"\033[38;2;136;136;136m{text}\033[0m"

# ── Rarity color themes ──
RARITY_THEMES = {
    "common":    {"border": rgb(158, 158, 158), "accent": rgb(189, 189, 189), "sprite": rgb(176, 176, 176)},
    "uncommon":  {"border": rgb(76, 175, 80),   "accent": rgb(129, 199, 132), "sprite": rgb(165, 214, 167)},
    "rare":      {"border": rgb(33, 150, 243),  "accent": rgb(100, 181, 246), "sprite": rgb(144, 202, 249)},
    "epic":      {"border": rgb(156, 39, 176),  "accent": rgb(206, 147, 216), "sprite": rgb(225, 190, 231)},
    "legendary": {"border": rgb(215, 119, 87),  "accent": rgb(255, 193, 7),   "sprite": rgb(255, 193, 7)},
}

RARITY_STARS = {
    "common": "★", "uncommon": "★★", "rare": "★★★", "epic": "★★★★", "legendary": "★★★★★"
}

ALL_SPECIES = ["all", "axolotl", "blob", "cactus", "capybara", "cat", "chonk", "dragon",
               "duck", "ghost", "goose", "mushroom", "octopus", "owl", "penguin", "rabbit",
               "robot", "snail", "turtle"]
ALL_RARITIES = ["all", "common", "uncommon", "rare", "epic", "legendary"]

# ── Terminal helpers ──
def hide_cursor():
    sys.stdout.write("\033[?25l")
    sys.stdout.flush()

def show_cursor():
    sys.stdout.write("\033[?25h")
    sys.stdout.flush()

def move_to(row, col):
    sys.stdout.write(f"\033[{row};{col}H")

def clear_screen():
    sys.stdout.write("\033[2J")
    move_to(1, 1)
    sys.stdout.flush()

def visual_width(s):
    """Count terminal columns — wide chars (emoji) count as 2."""
    w = 0
    for ch in s:
        if ch in ("✨", "💤", "●"):
            w += 2
        else:
            w += 1
    return w

# ── Load companions ──
def load_companions(core_only=False):
    CORE = {"anarathil", "eldurion", "faelindor", "ithrandur", "morgrath",
            "thistlewing", "thunderthistl", "vyrenth"}
    companions = []
    comp_dir = REPO / "companions"
    for buddy_path in sorted(comp_dir.rglob("buddy.json")):
        folder = buddy_path.parent.name
        if folder == "info":
            continue
        try:
            data = json.loads(buddy_path.read_text())
            bones = data["bones"]
            companions.append({
                "folder": folder,
                "name": data["name"],
                "personality": data.get("personality", ""),
                "species": bones["species"],
                "rarity": bones["rarity"],
                "eye": bones["eye"],
                "hat": bones["hat"],
                "shiny": bones.get("shiny", False),
                "stats": bones["stats"],
                "peakStat": data.get("peakStat", ""),
                "dumpStat": data.get("dumpStat", ""),
                "sprite_file": buddy_path.parent / "sprite.txt",
            })
        except (json.JSONDecodeError, KeyError):
            continue

    if core_only:
        companions = [c for c in companions if c["folder"] in CORE]
    return companions

def get_frame(sprite_file, frame_num, sleeping=False):
    """Extract a frame from sprite.txt."""
    if not sprite_file.exists():
        return [""] * 5
    text = sprite_file.read_text()
    if sleeping:
        label = "Sleeping"
    else:
        label = f"Frame {frame_num}"

    lines = []
    capturing = False
    for line in text.split("\n"):
        if f"--- {label}" in line:
            capturing = True
            continue
        if capturing:
            if line.startswith("---") or (line.strip() == "" and len(lines) >= 4):
                break
            lines.append(line)
    return (lines + [""] * 5)[:5]

# ── Stat bar ──
def stat_bar(val):
    filled = val // 5
    empty = 20 - filled
    return "█" * filled + "░" * empty

# ── Render card ──
def render_card(comp, frame_lines, sleeping, index, total, species_idx, rarity_idx, core_only):
    theme = RARITY_THEMES.get(comp["rarity"], RARITY_THEMES["legendary"])
    b = theme["border"]
    a = theme["accent"]
    s = theme["sprite"]
    cyan = rgb(130, 170, 220)
    green = rgb(147, 200, 130)

    hline = "─" * CARD_WIDTH
    blank = " " * CARD_WIDTH
    nav = f"{index + 1}/{total}"
    stars = RARITY_STARS.get(comp["rarity"], "?")
    species_upper = comp["species"].upper()
    rarity_upper = comp["rarity"].upper()
    shiny_tag = " ✨" if comp["shiny"] else ""

    lines = []

    # Top border
    lines.append(f"{LEFT_PAD}{b('╭' + hline + '╮')}")

    # Header
    header_left = f"{stars}  {rarity_upper}"
    gap = CARD_WIDTH - 2 - len(header_left) - len(species_upper)
    lines.append(f"{LEFT_PAD}{b('│')} {a(header_left)}{' ' * max(0, gap)}{cyan(species_upper)} {b('│')}{CLR}")

    # Blank
    lines.append(f"{LEFT_PAD}{b('│')}{blank}{b('│')}{CLR}")

    # Sprite (5 lines)
    for i in range(5):
        sprite_line = frame_lines[i] if i < len(frame_lines) else ""
        vw = visual_width(sprite_line)
        pad = max(0, CARD_WIDTH - 1 - vw)
        lines.append(f"{LEFT_PAD}{b('│')} {s(sprite_line)}{' ' * pad}{b('│')}{CLR}")

    # Blank
    lines.append(f"{LEFT_PAD}{b('│')}{blank}{b('│')}{CLR}")

    # Name
    name_display = f"{comp['name']}{shiny_tag}"
    vw = visual_width(name_display)
    pad = max(0, CARD_WIDTH - 1 - vw)
    lines.append(f"{LEFT_PAD}{b('│')} {bold_white(name_display)}{' ' * pad}{b('│')}{CLR}")

    # Blank
    lines.append(f"{LEFT_PAD}{b('│')}{blank}{b('│')}{CLR}")

    # Personality (3 lines word-wrapped)
    max_pw = CARD_WIDTH - 4
    words = comp["personality"].split()
    p_lines = []
    current = ""
    for word in words:
        if len(current) + len(word) + 1 > max_pw:
            p_lines.append(current)
            current = word
        else:
            current = f"{current} {word}" if current else word
    if current:
        p_lines.append(current)
    for pl in p_lines[:3]:
        pad = max(0, CARD_WIDTH - 1 - len(pl))
        lines.append(f"{LEFT_PAD}{b('│')} {dim_text(pl)}{' ' * pad}{b('│')}{CLR}")

    # Blank
    lines.append(f"{LEFT_PAD}{b('│')}{blank}{b('│')}{CLR}")

    # Stats
    for stat_name in ["DEBUGGING", "PATIENCE", "CHAOS", "WISDOM", "SNARK"]:
        val = comp["stats"].get(stat_name, 0)
        bar = stat_bar(val)
        suffix = ""
        if stat_name == comp["peakStat"]:
            suffix = " (peak)"
        elif stat_name == comp["dumpStat"]:
            suffix = " (dump)"
        vis_w = 2 + 10 + 1 + 20 + 1 + 3 + len(suffix) + 2
        right_pad = max(0, CARD_WIDTH - vis_w)
        stat_line = f"  {a(stat_name.ljust(10))} {bold_white(bar)} {dim_text(str(val).rjust(3))}{dim_text(suffix)}{' ' * right_pad}  "
        lines.append(f"{LEFT_PAD}{b('│')}{stat_line}{b('│')}{CLR}")

    # Blank
    lines.append(f"{LEFT_PAD}{b('│')}{blank}{b('│')}{CLR}")

    # Status
    if sleeping:
        pad = max(0, CARD_WIDTH - 17)
        lines.append(f"{LEFT_PAD}{b('│')}  {dim_text('💤 sleeping...')}{' ' * pad}{b('│')}{CLR}")
    else:
        pad = max(0, CARD_WIDTH - 11)
        lines.append(f"{LEFT_PAD}{b('│')}  {green('● online')}{' ' * pad}{b('│')}{CLR}")

    # Bottom border
    lines.append(f"{LEFT_PAD}{b('╰' + hline + '╯')}")
    lines.append("")

    # Footer
    lines.append(f"{LEFT_PAD}  {dim_text('a/← prev')}  {bold_white(f'[ {nav} ]')}  {dim_text('d/→ next')}  {dim_text('↑↓ jump')}  {dim_text('x exit')}{CLR}")
    if not core_only:
        sp = ALL_SPECIES[species_idx]
        ra = ALL_RARITIES[rarity_idx]
        lines.append(f"{LEFT_PAD}  {dim_text('Species:')} {cyan(sp)} {dim_text('(w/s)')}  {dim_text('Rarity:')} {cyan(ra)} {dim_text('(q/e)')}  {dim_text(comp['folder'])}{CLR}")

    return "\r\n".join(lines)

# ── Input ──
def get_key(timeout):
    """Non-blocking key read with timeout."""
    if select.select([sys.stdin], [], [], timeout)[0]:
        ch = sys.stdin.read(1)
        if ch == "\033":
            if select.select([sys.stdin], [], [], 0.05)[0]:
                ch += sys.stdin.read(2)
        return ch
    return None

# ── Main ──
def main():
    args = sys.argv[1:]
    core_only = "--core" in args
    delay = FRAME_DELAY
    start_at = ""

    for i, arg in enumerate(args):
        if arg == "--delay" and i + 1 < len(args):
            delay = float(args[i + 1])
        elif arg == "--start" and i + 1 < len(args):
            start_at = args[i + 1]

    print("Loading companions...")
    companions = load_companions(core_only)
    print(f"Loaded {len(companions)} companions.")

    if not companions:
        print("No companions found!")
        return

    # Filters
    species_idx = 0
    rarity_idx = 0

    def build_filtered():
        sp = ALL_SPECIES[species_idx]
        ra = ALL_RARITIES[rarity_idx]
        return [c for c in companions
                if (sp == "all" or c["species"] == sp)
                and (ra == "all" or c["rarity"] == ra)]

    filtered = build_filtered()
    index = 0

    # Find start companion
    if start_at:
        for i, c in enumerate(filtered):
            if c["folder"] == start_at:
                index = i
                break

    # Terminal setup
    old_settings = termios.tcgetattr(sys.stdin)
    try:
        tty.setraw(sys.stdin.fileno())
        hide_cursor()
        clear_screen()

        frame = 0
        need_reload = True

        while True:
            if need_reload:
                filtered = build_filtered()
                if index >= len(filtered):
                    index = 0
                if index < 0:
                    index = len(filtered) - 1
                need_reload = False
                frame = 0

            if not filtered:
                move_to(1, 1)
                orange = rgb(215, 119, 87)
                hline = "─" * CARD_WIDTH
                blank = " " * CARD_WIDTH
                sys.stdout.write(f"{LEFT_PAD}{orange('╭' + hline + '╮')}\r\n")
                sys.stdout.write(f"{LEFT_PAD}{orange('│')}{blank}{orange('│')}\r\n")
                msg = "No companions match filters"
                p = (CARD_WIDTH - len(msg)) // 2
                sys.stdout.write(f"{LEFT_PAD}{orange('│')}{' ' * p}{rgb(255,107,128)(msg)}{' ' * (CARD_WIDTH - p - len(msg))}{orange('│')}\r\n")
                sys.stdout.write(f"{LEFT_PAD}{orange('│')}{blank}{orange('│')}\r\n")
                sys.stdout.write(f"{LEFT_PAD}{orange('╰' + hline + '╯')}\r\n")
                sys.stdout.flush()
            else:
                comp = filtered[index]

                # Get frame
                if frame in (0, 1, 3, 4):
                    fl = get_frame(comp["sprite_file"], frame % 2, False)
                    sleeping = False
                elif frame == 2:
                    fl = get_frame(comp["sprite_file"], 2, False)
                    sleeping = False
                else:  # frame == 5
                    fl = get_frame(comp["sprite_file"], 0, True)
                    sleeping = True

                move_to(1, 1)
                output = render_card(comp, fl, sleeping, index, len(filtered),
                                     species_idx, rarity_idx, core_only)
                sys.stdout.write(output)
                sys.stdout.flush()

            # Input
            key = get_key(delay)
            if key:
                if key in ("x", "X"):
                    break
                elif key in ("\033[D", "a", "A"):  # Left
                    index = (index - 1) % max(1, len(filtered))
                    need_reload = True
                elif key in ("\033[C", "d", "D"):  # Right
                    index = (index + 1) % max(1, len(filtered))
                    need_reload = True
                elif key == "\033[A":  # Up
                    index = (index - 10) % max(1, len(filtered))
                    need_reload = True
                elif key == "\033[B":  # Down
                    index = (index + 10) % max(1, len(filtered))
                    need_reload = True
                elif key in ("e", "E") and not core_only:
                    rarity_idx = (rarity_idx + 1) % len(ALL_RARITIES)
                    index = 0
                    need_reload = True
                elif key == "q" and not core_only:
                    rarity_idx = (rarity_idx - 1) % len(ALL_RARITIES)
                    index = 0
                    need_reload = True
                elif key in ("s", "S") and not core_only:
                    species_idx = (species_idx + 1) % len(ALL_SPECIES)
                    index = 0
                    need_reload = True
                elif key in ("w", "W") and not core_only:
                    species_idx = (species_idx - 1) % len(ALL_SPECIES)
                    index = 0
                    need_reload = True

            frame = (frame + 1) % 6

    finally:
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_settings)
        show_cursor()
        clear_screen()
        comp_name = filtered[index]["name"] if filtered else "companion"
        print(f"Goodbye from {comp_name}!")

if __name__ == "__main__":
    main()
